### ISSUE TITLE
Problems with Logout - is the example working?

### ISSUE THREAD

**Issue by enjoyinglifeasadev** [IdentityServer/IdentityServer3#1228]
I have successfully created a sample identity server and separate MVC client using the information at http://identityserver.github.io/Documentation/docs/overview/mvcGettingStarted.html.  Everything I need is currently working except the redirect after a logout.

As a sanity check I have downloaded and run the "MVC Authentication" sample (https://github.com/IdentityServer/IdentityServer3.Samples/tree/master/source/MVC%20Authentication):

(1) I get a null reference exception on line 89 of startup.cs when I attempt to login.  This was fixed with the following code:
```
if (userInfo.Claims != null)
{
    userInfo.Claims.ToList().ForEach(ui => nid.AddClaim(new Claim(ui.Item1, ui.Item2)));
}
```

(2) With the above code fix in place, I now get a null reference exception when I log out.  Commenting out the `RedirectToIdentityProvider` handler prevents the exception but doesn't provide the redirect (for obvious reasons).  The stack trace doesn't show the sample code at all...
```
at System.Security.Claims.Claim..ctor(Claim other, ClaimsIdentity subject)
   at System.Security.Claims.Claim.Clone(ClaimsIdentity identity)
   at System.Security.Claims.ClaimsIdentity.SafeAddClaims(IEnumerable`1 claims)
   at System.Security.Claims.ClaimsIdentity..ctor(IIdentity identity, IEnumerable`1 claims, String authenticationType, String nameType, String roleType, Boolean checkAuthType)
   at System.Security.Claims.ClaimsIdentity..ctor(IEnumerable`1 claims, String authenticationType)
   at Thinktecture.IdentityModel.Identity.Create(String authenticationType, Claim[] claims) in c:\etc\Dropbox\thinktecture\IdentityModel\source\Core\Identity.cs:line 30
```

I believe that issue (2) is responsible for the problem that I am having with my own implementation of a split server and client.

N.B.  I am running this against a preview of VS2015, I very much doubt this is the problem but would be keen to know if there are any problems with the release.

**Comment by huysentruitw** [comment]
Try other samples from the Clients sub-directory first. Don't forget to run a IdSrv first (f.e. SelfHost (Minimal) )

**Comment by enjoyinglifeasadev** [comment]
Thanks for the prompt reply.  There is only one more sample that uses `RedirectToIdentityProvider` - 'MVC OWIN Client (Hybrid)'.  With the 'SelfHost (Minimal)' running, when I try to login to the client I get a very similar error on line 79 of startup.cs:

```
n.AuthenticationTicket = new AuthenticationTicket(new ClaimsIdentity(claims.Distinct(new ClaimComparer()), n.AuthenticationTicket.Identity.AuthenticationType), n.AuthenticationTicket.Properties);
```

The problem is with `new ClaimsIdentity(claims.Distinct(new ClaimComparer()), n.AuthenticationTicket.Identity.AuthenticationType)` as shown in the YSOD:

```
[NullReferenceException: Object reference not set to an instance of an object.]
   System.Security.Claims.Claim..ctor(Claim other, ClaimsIdentity subject) +11675562
   System.Security.Claims.Claim.Clone(ClaimsIdentity identity) +30
   System.Security.Claims.ClaimsIdentity.SafeAddClaims(IEnumerable`1 claims) +80
   System.Security.Claims.ClaimsIdentity..ctor(IIdentity identity, IEnumerable`1 claims, String authenticationType, String nameType, String roleType, Boolean checkAuthType) +11528894
   System.Security.Claims.ClaimsIdentity..ctor(IEnumerable`1 claims, String authenticationType) +21
   MVC_OWIN_Client.<<Configuration>b__2>d__0.MoveNext() in C:\Users\enjoyinglifeasadev\Documents\Visual Studio 2015\Projects\Thinktecture Samples\Clients\MVC OWIN Client (Hybrid)\Startup.cs:79
   System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task) +92
...
```

I will try and locate a machine with an older version of VS to rule this out. I'm curious to know if the examples are working for others...

**Comment by huysentruitw** [comment]
I tested the same combination (both on master and dev branch) without modifications and they work (VS2013).

![image](https://cloud.githubusercontent.com/assets/4618744/7251071/05161ff6-e828-11e4-9ce7-eca27c648815.png)

**Comment by enjoyinglifeasadev** [comment]
Must be an issue with VS2015, or possibly Windows 10 Preview (something I forgot earlier).  Have just performed the same steps as I did earlier and it ran first time.  Thanks for the pointers about 'Mvc Authentication' (not) being maintained.  I did manage to get this working with no modifications but I will be wary of it next time round.

Thanks for your time, and I will be a little more cautious about trial versions of VS and Windows in the future.
